tests/gamma/affinity_test.py

```diff
@@ -171,5 +175,121 @@ def test_session_affinity_policy_with_service_target(self):
             )
 
 
+class SessionDrainAffinityTest(xds_gamma_testcase.GammaXdsKubernetesTestCase):
+    def getClientRpcStats(
+        self,
+        test_client: _XdsTestClient,
+        num_rpcs: int,
+        *,
+        metadata_keys: Optional[tuple[str, ...]] = None,
+    ) -> grpc_testing.LoadBalancerStatsResponse:
+        """Load all metadata_keys by default."""
+        return super().getClientRpcStats(
+            test_client,
+            num_rpcs,
+            metadata_keys=metadata_keys or client_app.REQ_LB_STATS_METADATA_ALL,
+        )
+
+    def setUp(self):
+        self.pre_stop_hook = True
+        self.termination_grace_period_seconds = (
+            _TERMINATION_GRACE_PERIOD_SECONDS
+        )
+        super(SessionDrainAffinityTest, self).setUp()
+
+    def test_session_drain(self):
+        with self.subTest("01_run_test_server"):
+            test_servers = self.startTestServers(replica_count=_REPLICA_COUNT)
+
+        with self.subTest("02_create_ssa_policy"):
+            self.server_runner.createSessionAffinityPolicy(
+                "gamma/session_affinity_policy_route.yaml"
+            )
+
+        with self.subTest("03_create_backend_policy"):
+            self.server_runner.createBackendPolicy()
+
+        # Default is round robin LB policy.
+
+        with self.subTest("04_start_test_client"):
+            test_client: _XdsTestClient = self.startTestClient(test_servers[0])
+
+        with self.subTest("05_send_first_RPC_and_retrieve_cookie"):
+            (
+                cookie,
+                chosen_server,
+            ) = session_affinity_util.assert_eventually_retrieve_cookie_and_server(
+                self, test_client, test_servers
+            )
```

Comment:
https://github.com/grpc/grpc/pull/34519#discussion_r1340627434

nit:  hard to read - looks too much like lisp :)